Add API to the GC runtime

jacob-hughes opened the pull request:
I've split this up to be reviewed separately because I think our entry points into the GC from the standard library, user code, and our modified LLVM IR, are going to be fundamental in how the rest of the collector looks. 

I have created a collector as a static, mutable singleton struct which contains a bunch of bookkeeping information about the heap and rootset, and will contain all the methods needed for use within the GC runtime. 

In this draft PR, I'd like to focus solely on the interaction with the runtime externally, so I've not included `collector.rs` as I don't want its implementation details to leak into how we think about the higher level API. However, if you think it might be useful and help give you a bigger picture, let me know, and I'll push it.

Commit:
Flesh out collector API

This introduces a basic API for mutator threads to interact with the GC
runtime. Collection is managed through a thread local `Collector`
object. A lot of the functionality is unimplemented but this presents a
high-level overview of how the structure of the GC will look.

ltratt commented on src/api.rs:
```diff
@@ -0,0 +1,90 @@
+///     2. Allocate a chunk of heap memory to be used to store objects managed
+///        by the GC.
+pub fn init() {
+        COLLECTOR.alloc_heap().make_root_table();
```
I think this file hasn't gone through `rustfmt` ;)

jacob-hughes replied:
Guilty

ltratt commented on src/api.rs:
```diff
@@ -0,0 +1,90 @@
+//! This file contains the API for the runtime GC. Interaction with the GC is
+//! only legal through these functions. 
+//!
+//! The GC runtime consists of a mutable singleton `Collector` struct: 
```
I'm not saying we shouldn't do this, but I couldn't immediately work out how we might at some point generalise this to multiple threads. I wonder if the `Collector` struct should appear immutable but have `UnsafeCell`s inside?

jacob-hughes replied:
Perhaps you're right. I did worry about the current approach as it means any method on the collector with a `&mut self` would need to be placed in an unsafe block.

Using an `UnsafeCell` inside a static singleton would mean that the collector would need to be instantiated locally, (inside a `thread_local!` macro). I _think_ this sounds right to me. There will need to be atomic access to the collector, and this will mean that only the collector thread can mutate it.

The Rust code for this isn't very pretty, it would look a bit like this in places:

```rust
thread_local!(static GC: UnsafeCell<usize> = UnsafeCell::new(5));

fn main() {
    // Print the value
    GC.with(|f| {
        println!("{:?}", unsafe {*f.get()})
    });

    // mutate some GC value,
    // Then print the value
    GC.with(|f| {
        unsafe { *f.get() = 10;}
        println!("{:?}", unsafe {*f.get()})
    });
}
```

But at least we'd get finer grain control over where our unsafe boundaries are.

ltratt replied:
What's `with` do?

vext01 replied:
It grants safe access to the thread local:
> The with method yields a reference to the contained value which cannot be sent across threads or escape the given closure.

https://doc.rust-lang.org/std/thread/struct.LocalKey.html

ltratt replied:
`GC` has type `UnsafeCell<...>` so I can't see a `with` function. Maybe I'm being an idiot!

vext01 replied:
`thread_local!` wraps the value in a `LocalKey`.

> This key uses the fastest possible implementation available to it for the target platform. It is instantiated with the thread_local! macro and the primary method is the with method.

ltratt replied:
Aha, I see! I am an idiot :0

vext01 commented on src/api.rs:
```diff
@@ -0,0 +1,90 @@
+/// | We need to be really careful about code which panics inside the collector|
+/// | here. The safepoint poll will *not* be called by native Rust code which  |
+/// | means that panic handling is UB. We should probably abort on panic here. |
+/// ----------------------------------------------------------------------------
```
Should we use the `abort` panic strategy then?

ltratt replied:
This is tricky: as a library, our GC can't really impose panic or unwind on users. Ideally, we have to deal with whatever they select. I agree that unwind feels a bit of a mess in Rust.

jacob-hughes replied:
I'm not sure I have a good answer to this one. At the very least, when debugging, we'll want abort on panic so that we can have some idea what went wrong. But yes, for the user we would need to hopefully recover in some way.

vext01 replied:
Can the user really recover if the GC fails?

ltratt replied:
Yes. You can sometimes free up inessential resources (e.g. caches). It doesn't happen often though. Anyway, that's an orthogonal issue: in the long run we *have* to support unwinding, even if we don't like it very much.

vext01 replied:
I'm sceptical, but let's see what happens.

vext01 commented on src/api.rs:
```diff
@@ -0,0 +1,90 @@
+/// ----------------------------------------------------------------------------
+#[no_mangle]
+extern pub fn safepoint_poll() {
+    if COLLECTOR.should_collect() {
```
I wonder why you separated `should_collect` from `collect`?

Why not have one function `collect`, which early returns if there's nothing to do?

ltratt replied:
I think "should_collect" is set at a (potentially) non-safepoint, so you can only do the actual collection when you're in a safepoint.

jacob-hughes replied:
Laurie's right, but also, `should_collect()` is a small, trivial flag check method so it can be inlined. Performing the check in `collect()` would potentially mean an additional function call for some particularly performance critical part of the code.

vext01 commented on src/api.rs:
```diff
@@ -0,0 +1,90 @@
+    COLLECTOR.collect_next();
+}
+
+
```
Laurie has already mentioned rustfmt ;)

vext01 commented:
A few superficial comments from me.

I have to say, the quality of the code comments in this PR are really good!

jacob-hughes commented:
> I have to say, the quality of the code comments in this PR are really good!

@vext01 Thanks! I think it's particularly important for this part of the API because each call assumes a bunch of invariants are upheld with no real way to know about this without writing a bunch of  "this is UB / this is fine" requirements in the comments.

jacob-hughes commented:
I have now changed the API to use a `thread_local!` to house the main `Collector` struct locally on the GC thread. 

There are also a couple of mechanical commits which simply move `src/api.rs` into `src/lib.rs`.

jacob-hughes commented:
I messed up the pushing commits here before they were ready so apologies if they're a bit hard to follow. In short, I've moved `src/api.rs` to `src/lib.rs` and also added `collector.rs` where I start to flesh out the basic functionality the collector is expected to have.

This is now ready to review.

vext01 commented:
Fine by me.

ltratt commented on src/collector.rs:
```diff
@@ -0,0 +1,85 @@
+const HALIGN: usize = 4;
+
+pub(crate) struct Collector {
+    hptr: UnsafeCell<*mut usize>,
```
Can we not make this `Cell` and use `set` and so on?

jacob-hughes replied:
Yes, I'm not sure why I left this one as `UnsafeCell` when changing the others to `Cell`. I've pushed a new commit to address this now.

ltratt commented on src/lib.rs:
```diff
@@ -1,4 +1,99 @@
+/// pointless - it won't make a collection happen any faster. It simply sets a
+/// flag in the collector which is checked in order to decide whether to make a
+/// collection.
+pub fn force_collect() {
```
I am not sure about `force_collect` -- because it doesn't force a collection AFAICS? Or does it block the current thread until a collection has occurred?

jacob-hughes replied:
I'm not sure it's the best name either. You're right, that it _doesn't_ force a collection immediately, but it means that the next time the mutator hits a safepoint, the collector will pause to reclaim memory. It won't block the current thread.

ltratt replied:
Right, so we need to change the name to something that doesn't suggest that it forces a GC. I'd suggest something like `request_collect_soon` which is a bit of a mouthful, but at least not easily misunderstood.

jacob-hughes replied:
I've changed this to `request_collect_soon`, but I don't really like it. I'm not sure what would be better though.

ltratt replied:
Maybe `queue_collect` or `schedule_collect` or similar?

ltratt commented on src/lib.rs:
```diff
@@ -1,4 +1,99 @@
+/// should not be called directly by the user. Instead, it is exposed so that
+/// the standard library can build a GC smart pointer to a managed object.
+///
+/// Allocation to the GC heap is cheap and consists of: a heap bounds check; the
```
I don't think we should define what an allocation consists of in the public API. For example, in a non-moving collector it probably won't be a pointer bump.

jacob-hughes replied:
Good point, this is implementation specific and shouldn't be public.

ltratt replied:
Don't forget this bit.

ltratt commented on src/collector.rs:
```diff
@@ -0,0 +1,77 @@
+    }
+
+    #[inline]
+    pub fn collect_next(&self) {
```
I'm not sure why, but this name doesn't really have the right sense to me. It's more `please_collect_soon` or similar!

jacob-hughes replied:
What `collect_next` really means is `this_will_collect_the_next_time_we_hit_a_safepoint_poll`. "Please" implies that there may be a reason that the future collection may be aborted/postponed, but at the moment, that's not the case.

jacob-hughes commented:
Ready for re-review. I've also changed `alloc_raw` to return a result type instead of a raw pointer. This way we can leave it up to the library user to decide what to do on issues such as running out of memory.

jacob-hughes commented:
[As discussed in person] I've changed the API so that `force_collect` actually blocks to collect when called.

jacob-hughes commented:
Ready for re-review

ltratt commented:
Ready to squash?

jacob-hughes commented:
Squashed
